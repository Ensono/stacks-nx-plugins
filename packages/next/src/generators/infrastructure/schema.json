{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "Ensono Stacks Next Infrastructure Generator",
    "description": "The @ensono-stacks/next:infrastructure will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "openTelemetry": {
            "type": "boolean",
            "description": "Auto-instrumentation for opentelemetry",
            "alias": "o",
            "$default": "false",
            "x-prompt": "Would you like auto-instrumentation for OpenTelemetry?"
        }
    },
    "required": ["project"]
}
