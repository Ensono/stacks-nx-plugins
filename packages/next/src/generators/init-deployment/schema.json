{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "InitDeployment",
    "title": "Ensono Stacks Next Init Deployment Generator",
    "description": "The @ensono-stacks/next:init-deployment will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "libraryName": {
            "type": "string",
            "description": "Name provided for generated helm chart library",
            "alias": "l",
            "default": "stacks-helm-chart",
            "x-prompt": "Would you like to specify a name for the generated helm chart library?"
        }
    },
    "required": ["project"]
}
