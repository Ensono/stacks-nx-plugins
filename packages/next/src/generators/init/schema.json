{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        }
    },
    "required": ["project"]
}
