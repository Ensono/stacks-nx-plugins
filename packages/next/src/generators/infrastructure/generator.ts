import { StacksConfigError } from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';
import { formatFiles } from './utils/format-files';

export default async function infrastructureGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project));
    } catch (error) {
        if (error instanceof StacksConfigError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:infrastructure --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks);
}
