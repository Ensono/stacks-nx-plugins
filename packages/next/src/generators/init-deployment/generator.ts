import {
    formatFiles,
    executedDependantGenerator,
    hasGeneratorExecutedForProject,
    StacksError,
} from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';

export default async function initDeploymentGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    if (
        !executedDependantGenerator(
            tree,
            'WorkspaceDeployment',
            options.project,
        )
    )
        return false;
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'NextInitDeployment',
            true,
        )
    )
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project, options));
    } catch (error) {
        if (error instanceof StacksError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:init-deployment --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(
            'libs',
            options.libraryName,
            'build',
            'helm',
            '**',
            '*.yaml',
        ),
    ]);

    return runTasksInSerial(...tasks);
}
