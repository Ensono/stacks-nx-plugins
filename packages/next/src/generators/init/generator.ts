import {
    formatFiles,
    formatFilesWithEslint,
    addCustomTestConfig,
    deploymentGeneratorMessage,
    hasGeneratorExecutedForProject,
    verifyPluginCanBeInstalled,
    tsMorphTree,
} from '@ensono-stacks/core';
import {
    GeneratorCallback,
    joinPathFragments,
    readProjectConfiguration,
    Tree,
    updateProjectConfiguration,
} from '@nx/devkit';
import { runTasksInSerial } from '@nx/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';
import path from 'path';

import { NextGeneratorSchema } from './schema';
import { addEslint } from './utils/eslint';
import { eslintFix } from './utils/eslint-fix';
import {
    addReactAxeConfigToApp,
    addReactAxeDependency,
} from './utils/react-axe';
import { addStorybook } from './utils/storybook';
import updateTsConfig from './utils/tsconfig';

export default async function initGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (hasGeneratorExecutedForProject(tree, options.project, 'NextInit'))
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    const update = { ...project };

    if (
        project.sourceRoot &&
        project.sourceRoot ===
            update.targets?.build?.configurations?.development?.outputPath
    ) {
        update.targets.build.configurations.development = {};
    }

    updateProjectConfiguration(tree, project.name, update);

    tasks.push(addEslint(tree, project.root), addStorybook(tree, options));

    const ciCoverageConfig = {
        ci: {
            collectCoverage: true,
            coverageReporters: ['text', 'html'],
            collectCoverageFrom: [
                './**/*.{js,jsx,ts,tsx}',
                './!**/.next/**',
                './!**/*.d.ts',
                './!**/*.config.*',
                './!**/_app.*',
            ],
            codeCoverage: true,
            ci: true,
        },
    };

    await addCustomTestConfig(
        tree,
        readProjectConfiguration(tree, options.project),
        project.name,
        ciCoverageConfig,
    );

    tasks.push(
        formatFilesWithEslint(options.project),
        addReactAxeDependency(tree),
    );

    // update tsconfig.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.json'),
        ['next.config.js'],
    );

    // update tsconfig.spec.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.spec.json'),
    );

    const morphTree = tsMorphTree(tree);

    console.info(chalk.green`Attempting to add configuration for react-axe`);
    addReactAxeConfigToApp(project, morphTree);
    console.info(
        chalk.green`continuing execution of next:init generator after attempting to add react-axe`,
    );
    eslintFix(project, tree);

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks, () =>
        deploymentGeneratorMessage(
            tree,
            `nx g @ensono-stacks/next:init-deployment --project ${options.project}`,
        ),
    );
}
