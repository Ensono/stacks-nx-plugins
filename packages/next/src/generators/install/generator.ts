import { formatFiles, getProjects, Tree } from '@nrwl/devkit';

import { NextGeneratorSchema } from './schema';
import { addEslint } from './utils/eslint';

export default async function installGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    const project = getProjects(tree).get(options.project);

    if (!project) {
        throw new Error(
            `Cannot find the ${options.project} project. Please double check the project name.`,
        );
    }

    addEslint(tree, project.sourceRoot);

    await formatFiles(tree);
}
