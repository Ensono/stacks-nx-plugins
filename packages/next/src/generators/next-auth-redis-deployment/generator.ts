import {
    formatFilesWithEslint,
    readStacksConfig,
    getResourceGroup,
    hasGeneratorExecutedForProject,
} from '@ensono-stacks/core';
import {
    readProjectConfiguration,
    Tree,
    logger,
    formatFiles,
    generateFiles,
    runTasksInSerial,
} from '@nx/devkit';
import path from 'path';

import { NextAuthRedisDeploymentGeneratorSchema } from './schema';
import {
    updateProjectJsonHelmUpgradeTarget,
    updateProjectJsonTerraformPlanTarget,
} from './utils/update-targets';

export default async function nextAuthRedisDeploymentGenerator(
    tree: Tree,
    options: NextAuthRedisDeploymentGeneratorSchema,
) {
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'NextAuthRedisDeployment',
        )
    )
        return false;

    const project = readProjectConfiguration(tree, options.project);
    const stacksConfig = readStacksConfig(tree);

    // add common project files
    generateFiles(tree, path.join(__dirname, 'files/common'), project.root, {
        projectName: project.name,
        nonprodNextAuthUrl: `${project.name}.${stacksConfig.domain.internal}`,
        prodNextAuthUrl: `${project.name}.${stacksConfig.domain.external}`,
    });

    // add cloud project files
    generateFiles(
        tree,
        path.join(__dirname, `files/${stacksConfig.cloud.platform}`),
        project.root,
        {
            projectName: project.name,
            nonProdResourceGroup: getResourceGroup(stacksConfig, 'nonprod'),
            prodResourceGroup: getResourceGroup(stacksConfig, 'prod'),
        },
    );

    // Update project.json
    updateProjectJsonHelmUpgradeTarget(project, tree);
    updateProjectJsonTerraformPlanTarget(project, tree);

    await formatFiles(tree);

    return runTasksInSerial(formatFilesWithEslint(options.project), () => {
        logger.warn(`Do not forget to update your .env.local environment variables with values.
`);
    });
}
