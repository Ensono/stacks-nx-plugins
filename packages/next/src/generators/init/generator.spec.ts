import { testUpdateStacksConfig } from '@ensono-stacks/core';
import { Tree, readJson, updateJson } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { applicationGenerator } from '@nrwl/next';
import { Schema as NextSchema } from '@nrwl/next/src/generators/application/schema';

import generator from './generator';
import { NextGeneratorSchema } from './schema';

describe('next install generator', () => {
    let tree: Tree;
    const options: NextGeneratorSchema = { project: 'next-app' };

    async function createNextApp(schema?: Partial<NextSchema>) {
        tree = createTreeWithEmptyWorkspace();
        await applicationGenerator(tree, {
            name: 'next-app',
            style: 'css',
            ...schema,
        });

        testUpdateStacksConfig(tree, options.project);
    }

    describe('Project config', () => {
        beforeEach(async () => {
            await createNextApp();
        });

        it('should add the ci config in the test command in the project.json', async () => {
            tree.write(
                'next-app/project.json',
                JSON.stringify({
                    targets: { test: {} },
                    name: 'next-app',
                    sourceRoot: '/next-app',
                }),
            );

            await generator(tree, options);

            const projectConfig = readJson(tree, 'next-app/project.json');

            expect(projectConfig.targets.test).toMatchObject(
                expect.objectContaining({
                    configurations: {
                        ci: {
                            ci: true,
                            collectCoverage: true,
                            coverageReporters: ['text', 'html'],
                            collectCoverageFrom: [
                                './**/*.{js,jsx,ts,tsx}',
                                './!**/.next/**',
                                './!**/*.d.ts',
                                './!**/*.config.*',
                                './!**/_app.*',
                            ],
                            codeCoverage: true,
                        },
                    },
                }),
            );
        });

        it('should update the ci config in the test command in the project.json if there is an existing custom test config', async () => {
            tree.write(
                'next-app/project.json',
                JSON.stringify({
                    targets: {
                        test: {
                            executor: '@nrwl/jest:jest',
                            outputs: ['{workspaceRoot}/coverage/{projectRoot}'],
                            options: {
                                jestConfig: '/next-app/jest.config.ts',
                                passWithNoTests: true,
                            },
                            configurations: {
                                ci: {
                                    ci: true,
                                },
                            },
                        },
                    },
                    name: 'next-app',
                    sourceRoot: '/next-app',
                }),
            );

            await generator(tree, options);

            const projectConfig = readJson(tree, 'next-app/project.json');

            expect(projectConfig.targets.test).toMatchObject(
                expect.objectContaining({
                    executor: '@nrwl/jest:jest',
                    outputs: ['{workspaceRoot}/coverage/{projectRoot}'],
                    options: {
                        jestConfig: '/next-app/jest.config.ts',
                        passWithNoTests: true,
                    },
                    configurations: {
                        ci: {
                            ci: true,
                            collectCoverage: true,
                            coverageReporters: ['text', 'html'],
                            collectCoverageFrom: [
                                './**/*.{js,jsx,ts,tsx}',
                                './!**/.next/**',
                                './!**/*.d.ts',
                                './!**/*.config.*',
                                './!**/_app.*',
                            ],
                            codeCoverage: true,
                        },
                    },
                }),
            );
        });

        it('should update nx.json and tag executed generator true', async () => {
            await createNextApp({});
            await generator(tree, { ...options });

            const nxJson = readJson(tree, 'nx.json');

            expect(
                nxJson.stacks.executedGenerators.project[
                    options.project
                ].includes('NextInit'),
            ).toBeTruthy();
            expect(
                nxJson.stacks.executedGenerators.project[
                    options.project
                ].includes('NextInit'),
            ).toBe(true);
        });

        it('should return false from method and exit generator if already executed', async () => {
            await createNextApp({});

            updateJson(tree, 'nx.json', nxJson => ({
                ...nxJson,
                stacks: {
                    ...nxJson.stacks,
                    executedGenerators: {
                        project: {
                            [options.project]: ['NextInit'],
                        },
                    },
                },
            }));

            const gen = await generator(tree, {
                ...options,
            });

            expect(gen).toBe(false);
        });
    });

    describe('eslint', () => {
        beforeEach(async () => {
            await createNextApp();
        });

        it('should install and configure react specific eslint', async () => {
            await generator(tree, options);

            const packageJson = readJson(tree, 'package.json');

            expect(Object.keys(packageJson.devDependencies)).toEqual(
                expect.arrayContaining([
                    'eslint-plugin-testing-library',
                    '@typescript-eslint/eslint-plugin',
                ]),
            );
        });

        it('should throw if project is not defined', async () => {
            await expect(
                generator(tree, {
                    ...options,
                    project: 'unknown',
                }),
            ).rejects.toThrowError("Cannot find configuration for 'unknown'");
        });

        it('should merge default files with an existing tsconfig file', async () => {
            const defaultConfig = {
                extends: '../../tsconfig.base.json',
                compilerOptions: {
                    jsx: 'preserve',
                    allowJs: true,
                    esModuleInterop: true,
                    allowSyntheticDefaultImports: true,
                    strict: false,
                    forceConsistentCasingInFileNames: true,
                    noEmit: true,
                    resolveJsonModule: true,
                    isolatedModules: true,
                    incremental: true,
                    types: ['jest', 'node'],
                },
                include: [
                    'src/**/*.ts',
                    'src/**/*.tsx',
                    'src/**/*.js',
                    'src/**/*.jsx',
                    'next-env.d.ts',
                ],
                exclude: ['node_modules', 'jest.config.ts'],
            };
            const defaultSpecConfig = {
                extends: './tsconfig.json',
                compilerOptions: {
                    outDir: '../../dist/out-tsc',
                    module: 'commonjs',
                    types: ['jest', 'node'],
                    jsx: 'react',
                },
                include: [
                    'jest.config.ts',
                    'src/**/*.test.ts',
                    'src/**/*.spec.ts',
                    'src/**/*.test.tsx',
                    'src/**/*.spec.tsx',
                    'src/**/*.test.js',
                    'src/**/*.spec.js',
                    'src/**/*.test.jsx',
                    'src/**/*.spec.jsx',
                    'src/**/*.d.ts',
                ],
            };

            tree.write('next-app/tsconfig.json', JSON.stringify(defaultConfig));
            tree.write(
                'next-app/tsconfig.spec.json',
                JSON.stringify(defaultSpecConfig),
            );

            await generator(tree, options);

            const tsconfig = readJson(tree, 'next-app/tsconfig.json');
            expect(tsconfig?.include).toContain('next.config.js');
            expect(tsconfig?.include).toContain('**/*.js');
            expect(tsconfig?.include).toContain('next-env.d.ts');

            const tsconfigSpec = readJson(tree, 'next-app/tsconfig.spec.json');
            expect(tsconfigSpec?.include).toContain('jest.config.ts');
            expect(tsconfigSpec?.include).toContain('**/*.spec.js');
        });

        it('should merge default files with an existing tsconfig file and a src folder', async () => {
            const defaultConfig = {
                extends: '../../tsconfig.base.json',
                compilerOptions: {
                    jsx: 'preserve',
                    allowJs: true,
                    esModuleInterop: true,
                    allowSyntheticDefaultImports: true,
                    strict: false,
                    forceConsistentCasingInFileNames: true,
                    noEmit: true,
                    resolveJsonModule: true,
                    isolatedModules: true,
                    incremental: true,
                    types: ['jest', 'node'],
                },
                include: [
                    'src/**/*.ts',
                    'src/**/*.tsx',
                    'src/**/*.js',
                    'src/**/*.jsx',
                    'next-env.d.ts',
                ],
                exclude: ['node_modules', 'jest.config.ts'],
            };

            tree.write('next-app/tsconfig.json', JSON.stringify(defaultConfig));
            tree.write(
                'next-app/tsconfig.spec.json',
                JSON.stringify(defaultConfig),
            );
            tree.write('next-app/src/data.json', '{}');

            await generator(tree, options);

            const tsconfig = readJson(tree, 'next-app/tsconfig.json');
            expect(tsconfig?.include).toContain('src/**/*.ts');
            expect(tsconfig?.include).toContain('src/**/*.tsx');
            expect(tsconfig?.include).toContain('src/**/*.js');
            expect(tsconfig?.include).toContain('src/**/*.jsx');
            expect(tsconfig?.include).toContain('next-env.d.ts');
            expect(tsconfig?.include).toContain('next.config.js');
        });

        it('should merge defaults with an existing eslintrc.json file', async () => {
            const defaultConfig = {
                plugins: ['@nrwl/nx'],
                overrides: [
                    {
                        files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
                        extends: ['airbnb/base'],
                        plugins: ['@nrwl/nx'],
                        rules: {
                            'dot-notation': 'off',
                        },
                    },
                ],
            };

            tree.write(
                'next-app/.eslintrc.json',
                JSON.stringify(defaultConfig),
            );

            await generator(tree, options);

            const rootConfig = readJson(tree, 'next-app/.eslintrc.json');

            expect(rootConfig).toMatchObject(
                expect.objectContaining({
                    plugins: ['@nrwl/nx'],
                    extends: expect.arrayContaining([
                        'plugin:@nrwl/nx/react-typescript',
                        'plugin:testing-library/react',
                        'plugin:@next/next/recommended',
                        'next/core-web-vitals',
                    ]),
                    overrides: expect.arrayContaining([
                        expect.objectContaining({
                            files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
                            excludedFiles: ['jest.config.ts'],
                            rules: expect.objectContaining({
                                'testing-library/await-async-utils': 'error',
                                'testing-library/await-async-query': 'error',
                                'testing-library/no-wait-for-side-effects':
                                    'error',
                                'testing-library/no-manual-cleanup': 'error',
                                'testing-library/prefer-explicit-assert':
                                    'warn',
                                'testing-library/prefer-presence-queries':
                                    'warn',
                                'testing-library/prefer-wait-for': 'error',
                                'testing-library/prefer-user-event': 'warn',
                                'testing-library/no-debug': 'off',
                            }),
                            parserOptions: {
                                project: ['next-app/tsconfig(.*)?.json'],
                            },
                        }),
                    ]),
                }),
            );
        });
    });
});
