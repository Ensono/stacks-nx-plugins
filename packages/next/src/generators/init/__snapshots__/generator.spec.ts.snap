// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`next install generator Generator Dependencies should match the snapshot for layout.tsx file: next-app/src/app/layout.tsx 1`] = `
"import './global.css';
import Axe from "../components/a11y/axe";

export const metadata = {
  title: 'Welcome to next-app',
  description: 'Generated by create-nx-workspace',
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body><Axe />{children}</body>
    </html>
  );
}
"
`;
